File: DynamicPrograming/csharp/NonAdjacentSum.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class NonAdjacentSum
{
    public static int MaxSum(IReadOnlyList<int> nums)
    {
        var memo = new Dictionary<int, int>();
        return Dfs(0, nums, memo);
    }

    private static int Dfs(int index, IReadOnlyList<int> nums, IDictionary<int, int> memo)
    {
        if (index >= nums.Count)
        {
            return 0;
        }

        if (memo.TryGetValue(index, out var cached))
        {
            return cached;
        }

        var include = nums[index] + Dfs(index + 2, nums, memo);
        var exclude = Dfs(index + 1, nums, memo);
        var best = include > exclude ? include : exclude;
        memo[index] = best;
        return best;
    }
}

File: DynamicPrograming/go/non_adjacent_sum.go
package dynamicprogramming

// NonAdjacentSum returns maximum sum of numbers with no adjacent picks.
func NonAdjacentSum(nums []int) int {
	memo := make(map[int]int)

	var dfs func(int) int
	dfs = func(i int) int {
		if i >= len(nums) {
			return 0
		}
		if val, ok := memo[i]; ok {
			return val
		}
		include := nums[i] + dfs(i+2)
		exclude := dfs(i + 1)
		if include > exclude {
			memo[i] = include
		} else {
			memo[i] = exclude
		}
		return memo[i]
	}

	return dfs(0)
}
